File: 2019.03.16/javascript/main.js

Node = function(val, left=null, right=null) {
    this.val = val;
    this.left = left;
    this.right = right;

};

Node.prototype =  function() {
    function serialize(node=this) {
        if (node === null) {
            return "null"
        }

        return `${node.val}-${node.serialize(node.left)}-${node.serialize(node.right)}`
    }

    function deserialize(s=null, vals=null) {
        if (s === null && vals === null) {
            throw "Both parameters are null."
        }

        if (vals === null && typeof(s) === "string") {
            vals = s.split('-');
        }

        val = vals.shift()
        if (val === "null") {
            return null
        }
        var node = new Node(val);
        node.left = node.deserialize(s=null, vals=vals);
        node.right = node.deserialize(s=null, vals=vals);

        return node;
    }

    return {
        serialize: serialize,
        deserialize: deserialize
    }
}();

var node = new Node('root', new Node('left', new Node('left.left')), new Node('right'))
var serial = node.serialize();
if (node.deserialize(node.serialize()).left.left.val === "left.left") {
    console.log("TEST PASSED")
}
else {
    console.log("TEST FAILED")
}

File: 2019.03.16/python/main.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    @staticmethod
    def serialize(node):
        if node is None:
            return "None"
        return node.val + "-" + Node.serialize(node.left) + "-" + Node.serialize(node.right)

    @staticmethod
    def deserialize(s=None, vals=None):
        if s is None and vals is None:
            raise Exception("Both parameters are None.")
        if vals is None and s:
            vals = s.split('-') # convert serialize string to array
        val = vals.pop(0) # get and remove the first index

        if val == "None":
            return None

        node = Node(val)
        node.left = Node.deserialize(vals=vals)
        node.right = Node.deserialize(vals=vals)

        return node
        
if __name__ == "__main__":
    node = Node('root', Node('left', Node('left.left')), Node('right'))
    assert Node.deserialize(Node.serialize(node)).left.left.val == "left.left", "TEST FAILED"
    print("TEST PASSED")
    